catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java

```diff
@@ -23,50 +23,88 @@
 
 package org.catrobat.catroid.visualplacement;
 
-import android.os.SystemClock;
 import android.view.MotionEvent;
 import android.widget.ImageView;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public class VisualPlacementTouchListener {
 
 	private Mode mode;
-	private float previousY;
+	private static final long DELAY_THRESHOLD = 30;
+	private static final float DISTANCE_THRESHOLD = 10.f * 10.f;
 	private float previousX;
-	private long lastTimeMove = 0;
+	private float previousY;
+	private long touchDownTimestamp = 0;
+	private TimeProvider timeProvider;
+	private List<TouchEventData> eventData = new ArrayList<>();
+
+	private final class TouchEventData {
+		private final long timeStamp;
+		private final float xCoordinate;
+		private final float yCoordinate;
+
+		private TouchEventData(long time, float x, float y) {
+			timeStamp = time;
+			xCoordinate = x;
+			yCoordinate = y;
+		}
+	}
 
 	private void setMode(Mode mode) {
 		this.mode = mode;
 	}
 
+	public VisualPlacementTouchListener(TimeProvider timeProvider) {
+		this.timeProvider = timeProvider;
+	}
+
 	public boolean onTouch(ImageView imageView, MotionEvent event, CoordinateInterface coordinateInterface) {
 		if (event.getPointerId(0) == 0) {
 			float currentX = event.getRawX();
 			float currentY = event.getRawY();
+			long currentTimestamp = timeProvider.elapsedRealtime();
 
 			switch (event.getAction()) {
 				case MotionEvent.ACTION_DOWN:
 					setMode(Mode.TAP);
 					previousX = currentX;
 					previousY = currentY;
-					lastTimeMove = SystemClock.elapsedRealtime();
+					touchDownTimestamp = currentTimestamp;
+					eventData.add(new TouchEventData(currentTimestamp, currentX, currentY));
 					break;
 				case MotionEvent.ACTION_CANCEL:
 				case MotionEvent.ACTION_UP:
 					if (mode == Mode.TAP) {
 						imageView.setX(event.getX() - (float) imageView.getWidth() / 2);
 						imageView.setY(event.getY() - (float) imageView.getHeight() / 2);
+					} else {
+						removeOldEvents(currentTimestamp);
+						if (!eventData.isEmpty()) {
+							TouchEventData oldestEntry = eventData.get(0);
+							float dX = previousX - oldestEntry.xCoordinate;
+							float dY = previousY - oldestEntry.yCoordinate;
+							double distance = dX * dX + dY * dY;
+							if (distance < DISTANCE_THRESHOLD && distance != 0) {
+								imageView.setX(imageView.getX() - dX);
+								imageView.setY(imageView.getY() - dY);
+							}
+						}
```

Comment:
this basically depends on (as described in my comment on Line 86) if you want the current or the previous, if you want the current then this doesn't even need an else because then we have to do this always before (possible) position correction.

Reply:
Let's go with the currentX/Y please.